log/logfmt_logger.go

```diff
@@ -18,16 +38,25 @@ func NewLogfmtLogger(w io.Writer) Logger {
 }
 
 func (l logfmtLogger) Log(keyvals ...interface{}) error {
-	// The Logger interface requires implementations to be safe for concurrent
-	// use by multiple goroutines. For this implementation that means making
-	// only one call to l.w.Write() for each call to Log. We first collect all
-	// of the bytes into b, and then call l.w.Write(b).
-	b, err := logfmt.MarshalKeyvals(keyvals...)
-	if err != nil {
+	enc := logfmtEncoderPool.Get().(*logfmtEncoder)
+	enc.Reset()
+	defer logfmtEncoderPool.Put(enc)
+
+	if err := enc.EncodeKeyvals(keyvals...); err != nil {
+		return err
+	}
+
+	// Add newline to the end of the buffer
+	if err := enc.EndRecord(); err != nil {
 		return err
 	}
-	b = append(b, '\n')
-	if _, err := l.w.Write(b); err != nil {
+
+	// The Logger interface requires implementations to be safe for concurrent
+	// use by multiple goroutines. For this implementation that means making
+	// only one call to l.w.Write() for each call to Log. We call Write ourself
+	// with buf.Bytes() instead of using buf.WriteTo(l.w) as WriteTo can call
+	// Write multiple times.
+	if _, err := l.w.Write(enc.buf.Bytes()); err != nil {
```

Comment:
The last sentence of this comment is incorrect. Please remove.

The current implementation for `bytes.Buffer.WriteTo` will only make one call to `Write`. We use `buf.Bytes` in favor of `buf.WriteTo` because `buf.Bytes` is faster. `buf.WriteTo` updates the read position of `buf` based on the number of bytes written, while `buf.Bytes` does not. Since we discard unread data in `buf` after returning from this method, we don't need to pay the overhead of updating its read position.